[MRELEASE-1078] Support version policy to read and use SCM tags and commits

nielsbasjes opened the pull request:
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MRELEASE-1078) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MJAVADOC-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MJAVADOC-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify -Prun-its` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).



This my implementation on creating the option to follow the ideas of [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) to automatically determine the next version.

This patch assumes the required changes for maven-scm are available:
- Jira issue [SCM-977](https://issues.apache.org/jira/browse/SCM-977) 
- Pull request https://github.com/apache/maven-scm/pull/135


Summary of the changes:
- Extended the VersionPolicyRequest to include the SCM information.
- New version policy that retrieves the ChangeLog from SCM (which includes the commit messages and the tags).
- An optional configuration string `projectVersionPolicyConfig` that contains the optional config for the version policy. NOTE: I have tried to make it a clean addition of the configuration for the maven-release-plugin but I have not been successful in adding the extra config schema to the plugin. So for now it is a `String` which (because it is XML) must be included as CDATA.
- I've added an integration test that shows it works and what the configuration looks like.


I'm really looking forward to your feedback on how I can improve this.

michael-o reviewed (commented):
I see a lot of changes which do not belong here, but are ultimately good to be applied. Can you move them into a separate PR? I'd be happy to review them.

nielsbasjes commented:
> I see a lot of changes which do not belong here, but are ultimately good to be applied. Can you move them into a separate PR? I'd be happy to review them.

https://issues.apache.org/jira/browse/MRELEASE-1096
https://github.com/apache/maven-release/pull/132

nielsbasjes commented:
Note:
- I have rebased this on top of https://github.com/apache/maven-release/pull/132 to make sure it all keeps working. So the diff looks a lot bigger than it is.
- This requires a maven-scm that contains: 
  - [SCM-977] Support for retrieving tags from the changelog. 
  - i.e. the maven-scm master branch since 2022-05-27
  - So the CI build will fail.
- Also includes a few minor fixes around logging (only in IT tests) which will become a problem given the logging changes in the current maven-scm master branch.

nielsbasjes commented:
I have a question for the reviewers (@michael-o): 
This is essentially a plugin for a plugin for maven.
Is there a better way to do the configuration of this "plugin for a plugin for maven" ?

michael-o commented:
> I have a question for the reviewers (@michael-o): This is essentially a plugin for a plugin for maven. Is there a better way to do the configuration of this "plugin for a plugin for maven" ?

What do you mean? Maven Release Plugin is a not a plugin on top of a plugin. Maven Release Plugin => Maven Release => Maven SCM API => Maven SCM Providers

nielsbasjes commented:
> > I have a question for the reviewers (@michael-o): This is essentially a plugin for a plugin for maven. Is there a better way to do the configuration of this "plugin for a plugin for maven" ?
> 
> What do you mean? Maven Release Plugin is a not a plugin on top of a plugin. Maven Release Plugin => Maven Release => Maven SCM API => Maven SCM Providers

The Conventional Commits part has specific configuration. That "next version calculator" is a plugin for the maven-release plugin.

michael-o commented:
> > > I have a question for the reviewers (@michael-o): This is essentially a plugin for a plugin for maven. Is there a better way to do the configuration of this "plugin for a plugin for maven" ?
> > 
> > 
> > What do you mean? Maven Release Plugin is a not a plugin on top of a plugin. Maven Release Plugin => Maven Release => Maven SCM API => Maven SCM Providers
> 
> The Conventional Commits part has specific configuration. That "next version calculator" is a plugin for the maven-release plugin.

I know see, it is not a plugin in Maven sense, since it is not a Maven plugin with mojos. You simply provide a new implementatino for an interface pluggable at runtime, no?

nielsbasjes commented:
> > The Conventional Commits part has specific configuration. That "next version calculator" is a plugin for the maven-release plugin.
> 
> I know see, it is not a plugin in Maven sense, since it is not a Maven plugin with mojos. 

Correct, this VersionPolicy is not a maven plugin, and as a consequence my editor (IntelliJ) does not support me in setting the right options.

> You simply provide a new implementation for an interface pluggable at runtime, no?

Yes, and for this implementation there can be implementation specific configuration. I would like to offer settings configuration in the pom.xml of the project at hand.
Right now (and this is the best I was able to get running) you can see in `maven-release-plugin/src/it/projects/prepare/ccsemver-policy/pom.xml` that the config for this VersionPolicy is a single string in projectVersionPolicyConfig which contains an XML as CDATA... which I don't really like.

My question is simply: Is there a better way to do this?

```
<configuration>
  <projectVersionPolicyId>CCSemVerVersionPolicy</projectVersionPolicyId>

  <!-- projectVersionPolicyConfig is an XML structure:  -->
  <!-- versionTag: A regex with 1 capture group that MUST extract the project.version from the SCM tag. -->
  <!-- minorRules: A list regexes that will be matched against all lines in each commit message since   -->
  <!--             the last tag. If matched the next version is at least a MINOR update.                -->
  <!-- majorRules: A list regexes that will be matched against all lines in each commit message since   -->
  <!--             the last tag. If matched the next version is at least a MAJOR update.                -->
  <!-- If a match is found the commit will trigger either a minor or major version increase             -->
  <!-- instead of only a patch increase.  -->
  <projectVersionPolicyConfig>
    <![CDATA[
    <cCSemverConfig>
      <versionTag>^v([0-9]+(?:\.[0-9]+(?:\.[0-9]+)?)?)$</versionTag>
      <majorRules>
        <majorRule>^[a-zA-Z]+!(?:\([a-zA-Z0-9_-]+\))?: .*$</majorRule>
        <majorRule>^BREAKING CHANGE:.*$</majorRule>
      </majorRules>
      <minorRules>
        <minorRule>^feat(?:\([a-zA-Z0-9_-]+\))?: .*$</minorRule>
      </minorRules>
    </cCSemverConfig>
    ]]>
  </projectVersionPolicyConfig>

</configuration>
```

michael-o commented:
There is an option to inject an `Xpp3Dom` object into the plugin config which in turn could be passed to your class.

nielsbasjes commented:
> There is an option to inject an `Xpp3Dom` object into the plugin config which in turn could be passed to your class.

Cool! Can you please point me towards this option on how to do this. Possibly even an example in another project? I would really like to have this in there.

michael-o commented:
Good question. I don't know an example from the top of my head, but you can search `maven-sources` multirepo for `Xpp3Dom`. I am sure this will do. In Maven Settings we use this trick to inject config for a server to Maven Wagon.

michael-o commented:
Since this requires SCM 2, I'd recommend to try out the current snapshot of SCM 2 on top of this master. I know, that there will be failures...

nielsbasjes commented:
I take the clean master branch and only change in the main pom.xml

-    <scmVersion>1.13.0</scmVersion>
+    <scmVersion>2.0.0-M2-SNAPSHOT</scmVersion>

Then I run `mvn clean verify`  (without the -Prun-its !!)

Then the Maven Release Manager fails with a lot of errors (70 failures).
I get the same results with both java 8, 11 and 17

When looking through the errors I see these recurring ones:

- DefaultScmRepositoryConfiguratorTest.testGetConfiguredRepository:72 check provider expected:<s[vn]> but was:<s[tub-provider]>
- Expected text value 'release-label' but was 'original-tag' 
- Expected text value 'scm:svn:file://localhost/tmp/scm-repo/tags/release-label' but was 'scm:svn:file://localhost/tmp/scm-repo/trunk/' 
- ClassCast class org.apache.maven.scm.provider.ScmProviderRepositoryStub cannot be cast to class org.apache.maven.scm.provider.ScmProviderRepositoryWithHost

@michael-o Are these the failures you were referring to? 
Are these caused by the scm change from 1.x to 2.x ? 
I consider these errors something that should be fixed outside of this specific pull request. Do you agree?

michael-o commented:
> I take the clean master branch and only change in the main pom.xml
> 
> ```
> -    <scmVersion>1.13.0</scmVersion>
> +    <scmVersion>2.0.0-M2-SNAPSHOT</scmVersion>
> ```
> 
> Then I run `mvn clean verify` (without the -Prun-its !!)
> 
> Then the Maven Release Manager fails with a lot of errors (70 failures). I get the same results with both java 8, 11 and 17
> 
> When looking through the errors I see these recurring ones:
> 
>     * DefaultScmRepositoryConfiguratorTest.testGetConfiguredRepository:72 check provider expected:<s[vn]> but was:<s[tub-provider]>
> 
>     * Expected text value 'release-label' but was 'original-tag'
> 
>     * Expected text value 'scm:svn:file://localhost/tmp/scm-repo/tags/release-label' but was 'scm:svn:file://localhost/tmp/scm-repo/trunk/'
> 
>     * ClassCast class org.apache.maven.scm.provider.ScmProviderRepositoryStub cannot be cast to class org.apache.maven.scm.provider.ScmProviderRepositoryWithHost
> 
> 
> @michael-o Are these the failures you were referring to?
Correct.

> Are these caused by the scm change from 1.x to 2.x ?
That is the question whether this is just a DI issue.

> I consider these errors something that should be fixed outside of this specific pull request. Do you agree?
Yes, absolutely.

nielsbasjes commented:
I checkout the master twice and did some debugging with both settings.
So far I have tracked the `expected:<s[vn]> but was:<s[tub-provider]>` down to the DefaultScmRepositoryConfigurator being constructed (`@Inject`) with two different implementations of the `ScmManager`: One gets `DefaultScmManager` (from Maven SCM) the other `ScmManagerStub` (test class of the maven release manager).

nielsbasjes commented:
At this point my guess (I do not fully understand many of the details of the injection system used here) is that the root cause is related to some of the changes of https://github.com/apache/maven-release/pull/118 as I see changes in the ScmManagerStub parameters (like Priority) and also the PlexusJUnit4TestCase.java changes settings around the ScmManagerStub. @cstamas does this make sense?

nielsbasjes commented:
> Good question. I don't know an example from the top of my head, but you can search `maven-sources` multirepo for `Xpp3Dom`. I am sure this will do. In Maven Settings we use this trick to inject config for a server to Maven Wagon. @rfscholte @hboutemy Do you know an example from the top of your head?

I have tried to figure out how this could work and I have not succeeded. So if you guys can point me to an example then that would be awesome. My main problem is that I do not understand how things are loaded. 

At this point I would put an extra method to obtain a `thing` that understands the `VersionPolicy` specific config and somehow make that a part of the system that reads the main config. Then when a specific policy is actually used the config is available.

Any pointers towards making this possible is highly appreciated.

michael-o commented:
@slawekjaranowski Can you help since you looked into XML-based mojo config for validation?

nielsbasjes commented:
How shall we proceed? Commit the way it is now (the config is an xml within a cdata block) or wait till the config has been changed to the way I would like it (where I am not yet able to do that)?
@michael-o / @slawekjaranowski ?

michael-o commented:
@nielsbasjes I thing I know for sure is Wagon config in `settings.xml` is `Xpp3Dom`. That is passed to Resolver down to Wagon. That could be a start.

nielsbasjes commented:
@michael-o 
- I have updated this by rebasing it on top of https://github.com/apache/maven-release/tree/maven-scm-2.0.0
- The `mvn -Prun-its clean verify` works on my machine (running on JDK 17).

I'm still trying to figure out how the config injection with the WagonConfigurator interface works.

nielsbasjes commented:
To summarize where I'm stuck with the 'clean' loading of the configuration.

I have been doing some more digging and I think what I would like is not possible (not like I had initially thought it up).

What I would like is to have as a config of this plugin (simplified):
```
<configuration>
   <projectVersionPolicyId>Foo</projectVersionPolicyId>
   <projectVersionPolicyConfig>
       <fooA>1</fooA>
       <fooB>true</fooB>
   </projectVersionPolicyConfig>
</configuration>
```
where this is part of the release plugin config
```
<configuration>
   <projectVersionPolicyId>Foo</projectVersionPolicyId>
   <projectVersionPolicyConfig>
...
   </projectVersionPolicyConfig>
</configuration>
```

and this is part of the `Foo` config
```
<fooA>1</fooA>
<fooB>true</fooB>
```

It is only possible to determine the `<fooA>` is a valid tag AFTER the `<projectVersionPolicyId>` has been read, the mentioned code (`Foo`) has been loaded and the possible config schema from there has been retrieved.
But in order to get to the `<projectVersionPolicyId>` the system needs to read (at least) the entire `<configuration>` block; which will fail over the (then) unknown `<fooA>` part.

I.e. Chicken and egg.

Based on this I now see only two ways to handle this:
1. As I have done now: The extra config is a CDATA string that is handled as I have written now.
2. As antrun does it: The config can be `any xml` which is not validated during the initial read.

I'm looking into the latter right now.

michael-o commented:
Let me first complete next SCM milestone, integrate into release and then tackle your problem. ok for you?

michael-o commented on maven-release-manager/src/main/mdo/release-descriptor.mdo:
```diff
@@ -491,10 +491,10 @@
           <name>projectVersionPolicyConfig</name>
           <version>3.0.0+</version>
-          <type>String</type>
+          <type>DOM</type>
```
DOM is exactly what you need. Just like here: https://github.com/apache/maven/blob/8125b3131f2e0f88837040b0baf4e0aaa8210679/maven-settings/src/main/mdo/settings.mdo#L574-L582 

 I assumed that I have provided this pointer to you.

michael-o commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/config/PropertiesReleaseDescriptorStore.java:
```diff
@@ -271,7 +271,7 @@ public void write( BuilderReleaseDescriptor config, File file )
         if ( config.getProjectVersionPolicyConfig() != null )
         {
-            properties.setProperty( "projectVersionPolicyConfig", config.getProjectVersionPolicyConfig() );
+            properties.setProperty( "projectVersionPolicyConfig", config.getProjectVersionPolicyConfig().toString() );
```
Why `toString()`? I don't understand this reasoning.

nielsbasjes replied:
Also: As a consequence of this (i.e. the config can possibly be written to a file and then read again) I assumed a String config and parse it in the VersionPolicy.

In the more generic usecase I like this because it leaves the format of the config fully to the writer of the VersionPolicy.

michael-o replied:
How to do you that `toString()` will give you something reasonable?

nielsbasjes replied:
I did some more check and currently I think is actually correct code.

What I have now:
- By setting the type to DOM the generated code makes the class of this thing an `Object` in Java (i.e. all getters and setters). I checks and the actuall class is an instance of org.codehaus.plexus.configuration.xml.XmlPlexusConfiguration. So the toString() recreates the XML fragment.
- Some of the code can only handle a `String` (like for example the parts that reads and writes the release.properties file).

So the code path now is roughly:
- The `ProjectVersionPolicyConfig` part in the pom.xml is parsed into a generic XmlPlexusConfiguration by Maven.
- This XmlPlexusConfiguration is converted into a `String` so it can be stored and processed by all intermediate components (some of which can only handle String).
- The actual VersionPolicy has it's own parser to make sense of the provided String.

I did two tests to verify this:

1) In the integration test I have put this in the pom.xml
```
<projectVersionPolicyConfig>
  <versionTag>^v([0-9]+(?:\.[0-9]+(?:\.[0-9]+)?)?)$</versionTag>
  <majorRules>
    <majorRule>^[a-zA-Z]+!(?:\([a-zA-Z0-9_-]+\))?: .*$</majorRule>
    <majorRule>^BREAKING CHANGE:.*$</majorRule>
  </majorRules>
  <minorRules>
    <minorRule>^feat(?:\([a-zA-Z0-9_-]+\))?: .*$</minorRule>
  </minorRules>
</projectVersionPolicyConfig>
```

I see this in the release.properties:
```
projectVersionPolicyConfig=<projectVersionPolicyConfig><versionTag>^v([0-9]+(?\:\\.[0-9]+(?\:\\.[0-9]+)?)?)$</versionTag>\n<majorRules><majorRule>^[a-zA-Z]+\!(?\:\\([a-zA-Z0-9_-]+\\))?\: .*$</majorRule>\n<majorRule>^BREAKING CHANGE\:.*$</majorRule>\n</majorRules>\n<minorRules><minorRule>^feat(?\:\\([a-zA-Z0-9_-]+\\))?\: .*$</minorRule>\n</minorRules>\n</projectVersionPolicyConfig>\n
```

2) To see if `any` config format is usable I put this in the pom.xml
```
<projectVersionPolicyConfig>
<![CDATA[
config:
  - something:
      foo:
        - 'foo foo'
      bar:
        - 'bar bar'
]]>
</projectVersionPolicyConfig>
```

I see this in the release.properties
```
projectVersionPolicyConfig=<projectVersionPolicyConfig>config\:\n  - something\:\n      foo\:\n        - 'foo foo'\n      bar\:\n        - 'bar bar'</projectVersionPolicyConfig>\n
```

In both these checks the `toString` made sure the right config is stored and also by `forcing` the VersionPolicy to have its own parser it works as I would like it to work.

This way it is up to the VersionPolicy to actually parse the config and verify what it is getting.

nielsbasjes commented:
@michael-o I've figured out how to do it and this is what the code now is.
I've started on some documentation, still a bit rough and very needed.

michael-o reviewed (commented):
Please look into `org.eclipse.aether.internal.transport.wagon.PlexusWagonConfigurator`. Here you'll find how to convert this DOM object into your class model. It does work from `settings.xml` to Wagon, it will work for you as wekk.

nielsbasjes commented:
I've done some testing (locally) with a real project and it all works as I would like it to work: https://github.com/nielsbasjes/yauaa/commit/a422037cb88bd899bdfcc4bc8c8c2af818fc9b52

nielsbasjes commented:
I've cleaned up the code and squashed it all into a single clean commit.
This does need the update to the new SCM version before it can be committed.

nielsbasjes commented:
Rebased against the master which now includes the needed SCM dependency.

michael-o commented:
I have worked recently on many reporting plugins which used `PlexusConfiguration`. I will get back to this end of month.

nielsbasjes commented:
@michael-o What can I do to get this moving?

michael-o commented:
> @michael-o What can I do to get this moving?

At moment nothing. This was too big to process for M7 for me personally. I wanted to upgrade to SCM 2 first. This will schedule for the next milestone. BTW, I have found the component which uses the DOM config. Need to recall from memory.

nielsbasjes commented:
My current patch already does that too.

nielsbasjes commented on maven-release-plugin/src/it/projects/prepare/ccsemver-policy/pom.xml:
```diff
@@ -0,0 +1,83 @@
+          <!--             the last tag. If matched the next version is at least a MAJOR update.                -->
+          <!-- If a match is found the commit will trigger either a minor or major version increase             -->
+          <!-- instead of only a patch increase.  -->
+          <projectVersionPolicyConfig>
+            <versionTag>^v([0-9]+(?:\.[0-9]+(?:\.[0-9]+)?)?)$</versionTag>
+            <majorRules>
+              <majorRule>^[a-zA-Z]+!(?:\([a-zA-Z0-9_-]+\))?: .*$</majorRule>
+              <majorRule>^BREAKING CHANGE:.*$</majorRule>
+            </majorRules>
+            <minorRules>
+              <minorRule>^feat(?:\([a-zA-Z0-9_-]+\))?: .*$</minorRule>
+            </minorRules>
+          </projectVersionPolicyConfig>
```
@michael-o This shows how the DOM based config looks now. This is the way I had in mind.

michael-o replied:
Yeah, let me first find the component I already forgot and then take care of this.

nielsbasjes replied:
@michael-o Do you have the additional info you mentioned a while ago?
Just let me know what I need to change to make it meet this projects requirements.

nielsbasjes commented:
CI Seems flaky today: 
> Connect to repo.maven.apache.org:443 [repo.maven.apache.org/146.75.28.215] failed: Connection timed out: connect -> [Help 1]

nielsbasjes commented:
Just did a rebase upon the latest master branch.

nielsbasjes commented:
@rfscholte @michael-o 
I just had a long talk with @hboutemy and we also discussed this merge request.

As an option that I'm fine with; how about I split this merge request into 2 parts:
1) Extending the VersionPolicy api with the capabilities needed to make a new class of versionplicies possible (among which the conventional commits).
2) I publish the conventional commits version policy that uses these new api capabilities under my own name (i.e. outside Apache).

This way the level of code you guys need to maintain is much smaller and thus this pull request becomes easier(?) for you to review.

Just let me know what you prefer. I'm fine either way.

michael-o commented:
> @rfscholte @michael-o I just had a long talk with @hboutemy and we also discussed this merge request.
> 
> As an option that I'm fine with; how about I split this merge request into 2 parts:
> 
>     1. Extending the VersionPolicy api with the capabilities needed to make a new class of versionplicies possible (among which the conventional commits).
> 
>     2. I publish the conventional commits version policy that uses these new api capabilities under my own name (i.e. outside Apache).
> 
> 
> This way the level of code you guys need to maintain is much smaller and thus this pull request becomes easier(?) for you to review.
> 
> Just let me know what you prefer. I'm fine either way.

This, I truly like. Extend and review API, then your implementation!

nielsbasjes commented:
Ok, then I'll change this in the coming weeks.

michael-o commented:
> Ok, then I'll change this in the coming weeks.

Dank U!

nielsbasjes commented:
Taking things out is much simpler than making things :)

I took out the version policy itself and the related documentation and comments.
I did leave some of the documentation in as it clarified what a VersionPolicy is.

This is the cleanest I could think of.

I have tested it by building my (out of project) VersionPolicy : https://github.com/nielsbasjes/conventional-commits-maven-release

nielsbasjes commented:
Done.
I think this is good now.
@michael-o WDYT?

nielsbasjes commented:
Note: This should really be squashed ...

michael-o reviewed (changes requested):
Since the actual implementation is now gone, the title of this PR as well as the JIRA issue should be renamed/rephrased. @nielsbasjes WDYT?

nielsbasjes commented:
@michael-o Do you like this one?
`Support for a VersionPolicy to use scm tags and commits.`

When merging this should be put in the squashed commit message also.

michael-o commented:
> @michael-o Do you like this one? `Support for a VersionPolicy to use scm tags and commits.`
> 
> When merging this should be put in the squashed commit message also.

This looks good, I'd fine to it: Support version policy to read and use SCM tags and commits

WDYT?

nielsbasjes commented:
Updated.

michael-o commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/config/ReleaseDescriptorBuilder.java:
```diff
@@ -401,6 +401,18 @@ public ReleaseDescriptorBuilder setProjectVersionPolicyId( String projectVersion
+    public ReleaseDescriptorBuilder setProjectVersionPolicyConfig( String setProjectVersionPolicyConfig )
+    {
+        releaseDescriptor.setProjectVersionPolicyConfig( setProjectVersionPolicyConfig );
+        return this;
```
This I don't understand. If the release descriptor supports an object, why not pass the original one instead of a string representation?

michael-o replied:
I know, but that conversion can be done when the properties object is populated, not earlier, no?

michael-o commented on maven-release-api/src/main/java/org/apache/maven/shared/release/config/ReleaseDescriptor.java:
```diff
@@ -426,6 +426,13 @@
+     * @return The parsed XML of the provided config (an instance of XmlPlexusConfiguration) or null.
+     */
+    Object getProjectVersionPolicyConfig();
+
```
Does not really make sense to tie it to XML, consider someone is using this outside of Maven realm and uses real objects throughout. Why not create a simple interface `VersionPolicyConfig` which has a `#serialize()` method. For the PlexusDom we can create an impl which will delegate to `#toString()`, thus we can cleanly separate between Maven Release and Maven Core usage. WDYT?`

nielsbasjes replied:
Given that this is a maven plugin I do not understand the `using this outside of Maven realm`.
The ProjectVersionConfig has been typed as a `DOM` and thus it is always an XML.
I find this a perfectly acceptable constraint for this kind of config.

nielsbasjes replied:
Also, I do not fully understand what you have in mind.

michael-o replied:
I don't agree with that. The Maven Release API is *not* the plugin. The Maven Release Plugin is a user of Maven Release. What if someone wants to write a custom plugin which uses the Maven Release API w/o our plugin?

nielsbasjes replied:
Ok, I'll have a look what I can do to improve this. I like the `VersionPolicyConfig` proposal.

nielsbasjes replied:
Trying to make it cleaner.
Observations so far:
- In the Modello of the `ModelloReleaseDescriptor` the config needs to be of type `DOM`. If it isn't then I have to wrap the config into CDATA which makes it a terrible developer experience.
- This generated class contains: `private Object projectVersionPolicyConfig;` and the related getter and setter that go with this also use `Object` as the class. 
- I checked the Modello code and I do not see a way to let it return a different type (like a `VersionPolicyConfig`).
- These getters and setters must be the same in the `ReleaseDescriptor` interface (i.e. also `Object`.

This is the code you mentioned.
If I simply change the return type then this results in a conflict with the generated code. I'm unable to change the generated code regarding this type.

Note that the other reason for the way I have it now was that at some point the config is also stored in a Property ... which only accepts String.

Please advise. I'm a bit stuck.

michael-o replied:
Alright, thanks for your investigation" Let me double check this tomorrow, maybe there is currently no better way to solve

nielsbasjes replied:
Have you been able to find a better way? 
I'm happy to implement it if it is possible.

michael-o replied:
Haven't looked, was busy with Resolver. Will dedicate some time to it this evening.

michael-o replied:
`org.apache.maven.shared.release.config.ModelloReleaseDescriptor` does config any general purpose config element and nothing as `DOM`. Can you clarify what you are referring to?

nielsbasjes replied:
In the `mdo` file you can specify a bunch of types and you can specify (xml) nesting.

If you cannot specify the entire thing in there (like in this case because all custom Version Policies will have currently unknown config options) you either end up with
- it being a `string` which means you have to CDATA the config which results in a very bad developer experience, but it does allow you to have any kind of config format you like (xml, json, yaml, ...) .
- it being a `dom` which makes it an XML without schema. The effect is that an IDE can only assist you in making sure it is an XML (which is not validated other than having the open and close tags in sync). And the effect is that the generate Modello code says it is an `Object` which causes what I tried to describe above.

michael-o replied:
Ah OK, I was confused by your statement because the modello model did not change. Can you update your findings in that regard and I will check again?

nielsbasjes replied:
I do not understand right now what you would like me to update.

michael-o replied:
I wonder if this is always going to be `XmlPlexusConfiguration` why not make it `XmlPlexusConfiguration`? Too much overhead? I mean the Javadoc is clear on this, no?

nielsbasjes replied:
I put in the JavaDoc it is really this type. 
It is the code generated by Modello that dictates to to be `Object`. 

I have tried changing it to `XmlPlexusConfiguration` but the current Modello code generator won't let me.

The only way to actually make it an `XmlPlexusConfiguration` is by changing (probably) this: https://github.com/codehaus-plexus/modello/blob/master/modello-plugins/modello-plugin-java/src/main/java/org/codehaus/modello/plugin/java/JavaModelloGenerator.java#L1368

michael-o replied:
Sigh, there is even an open TODO regarding this in Modello code. So, I guess we need to swallow the `Object`.

michael-o commented on maven-release-api/src/main/java/org/apache/maven/shared/release/versions/VersionParseException.java:
```diff
@@ -34,4 +34,14 @@ public VersionParseException( String message )
+     *
+     * @param message a {@link java.lang.String} object
+     */
+    public VersionParseException( String message, Exception e )
```
`Throwable` like the superclass?

nielsbasjes replied:
Yes

michael-o commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/config/ReleaseDescriptorBuilder.java:
```diff
@@ -401,6 +401,18 @@ public ReleaseDescriptorBuilder setProjectVersionPolicyId( String projectVersion
+     */
+    public ReleaseDescriptorBuilder setProjectVersionPolicyConfig( String setProjectVersionPolicyConfig )
+    {
+        releaseDescriptor.setProjectVersionPolicyConfig( setProjectVersionPolicyConfig );
```
So if this would be the `XmlPlexusConfiguration` you'd be force to parse the serialized config into that type?

nielsbasjes replied:
No, it is an XML. But other than that you can parse it any way you like.
See for example my current implementation that uses Jackson.
https://github.com/nielsbasjes/conventional-commits-maven-release/blob/main/src/main/java/nl/basjes/maven/release/version/conventionalcommits/ConventionalCommitsVersionConfig.java

michael-o replied:
I see, but your `toString()` does not spit out XML: https://github.com/nielsbasjes/conventional-commits-maven-release/blob/56d1e01481150ebb2e3f5afea111effcdfa25b74/src/main/java/nl/basjes/maven/release/version/conventionalcommits/ConventionalCommitsVersionConfig.java#L122-L129 since you use it at a higher layer, right?

michael-o commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/config/ReleaseUtils.java:
```diff
@@ -156,6 +156,10 @@ public static void copyPropertiesToReleaseDescriptor( Properties properties, Rel
         }
+        if ( properties.containsKey( "projectVersionPolicyConfig" ) )
+        {
+            builder.setProjectVersionPolicyConfig( properties.getProperty( "projectVersionPolicyConfig" ) );
```
Here as well, I guess?!

michael-o commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/phase/AbstractMapVersionsPhase.java:
```diff
@@ -67,9 +75,16 @@
  * @author Robert Scholte
  */
+@Component( role = ReleasePhase.class, hint = "map-release-versions" )
 public abstract class AbstractMapVersionsPhase
```
This is then inherited to the non-abstract class?

nielsbasjes replied:
Yes, it seems that way. If it wasn't then the scmRepositoryConfigurator would have remained null causing an NPE.

michael-o replied:
I know the reason and I believe this is wrong. We have changed to constructor injection. Thus the implementing classes have `@Inject` annotated constructors and there your instance should be provided. @cstamas

nielsbasjes replied:
If I remove this like then I get an NPE as expected.

michael-o replied:
Even with constructor injection? @cstamas Can you help? I have intentionally moved from Plexus annotations to standard Java annotations

nielsbasjes replied:
Ah, yes. I missed you mentioned @Inject.
Fixing it.

nielsbasjes replied:
Fixed.

nielsbasjes replied:
Fixed.

michael-o commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/phase/AbstractMapVersionsPhase.java:
```diff
@@ -347,16 +362,42 @@ private String getContextString( ReleaseDescriptor releaseDescriptor )
+            try
+            {
+                ScmRepository repository = scmRepositoryConfigurator
+                        .getConfiguredRepository( releaseDescriptor, new Settings() );
```
Settings don't matter if they are requied?

nielsbasjes replied:
Good point. Fixing it.

nielsbasjes replied:
Fixed

michael-o commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/phase/AbstractMapVersionsPhase.java:
```diff
@@ -347,16 +362,42 @@ private String getContextString( ReleaseDescriptor releaseDescriptor )
+            }
+            catch ( ScmRepositoryException | NoSuchScmProviderException e )
+            {
+                getLogger().warn( "Next Version will NOT be based on the version control: {}", e.getMessage() );
```
This loses stack trace, we try to apply the following scheme:

```
if (log.isWarn())
getLogger().warn( "Next Version will NOT be based on the version control")
else if (log.isDebug())
getLogger().warn( "Next Version will NOT be based on the version control", e)
```

nielsbasjes replied:
Fixed

nielsbasjes commented:
Thanks everyone.
